lib/puppet/catalog-diff/compilecatalog.rb

```diff
@@ -95,23 +97,25 @@ def compile_catalog(node_name, server, certless)
         endpoint = "/puppet/v3/catalog/#{node_name}?environment=#{environment}"
       end
 
-      Puppet.debug("Connecting to server: #{server}#{endpoint}")
+      uri = URI("https://#{server}:#{port}#{endpoint}")
+      Puppet.debug("Connecting to server: #{uri}")
       begin
-        connection = Puppet::Network::HttpPool.http_instance(server, port)
-
         ret = if certless
-                connection.request_post(endpoint, body.to_json, headers).body
+                Puppet.runtime[:http].post(uri, body.to_json, headers: headers)
               else
-                connection.request_get(endpoint, headers).body
+                Puppet.runtime[:http].get(uri, headers: headers)
               end
+      unless ret.success?
+        raise "HTTP request to PuppetDB failed with: HTTP #{ret.code} - #{ret.reason}"
+      end
       rescue Exception => e
         raise "Failed to retrieve catalog for #{node_name} from #{server} in environment #{environment}: #{e.message}"
       end
 
       begin
-        catalog = PSON.parse(ret)
+        catalog = PSON.parse(ret.body)
       rescue PSON::ParserError => e
-        raise "Error parsing json output of puppet catalog query for #{node_name}: #{e.message}. Content: #{ret}"
+        raise "Error parsing json output of puppet catalog query for #{node_name}: #{e.message}. Content: #{ret.body}"
```

Comment:
Just of a general comment, unrelated to this PR. Puppet switched from PSON to JSON by default in Puppet 5, we'd like to drop PSON at some point (https://tickets.puppetlabs.com/browse/PUP-9648), and PSON is not suitable to encoding rich data (https://tickets.puppetlabs.com/browse/PUP-10928). Alternatively, you could simplify this code to "get a catalog the same way the agent does without using the indirector". Something like:

```ruby
http = Puppet.runtime[:http]
session = http.create_session
compiler = session.route_to(:compiler)
catalog = compiler.post_catalog(...)
```

Here is documentation about the arguments to [`post_catalog`](https://www.rubydoc.info/gems/puppet/Puppet/HTTP/Service/Compiler#post_catalog-instance_method)

Also this is how the agent calls the same method to [get a catalog]( https://github.com/puppetlabs/puppet/blob/ac45fe2b783708f64cb3a28f21ee92673e1f3216/lib/puppet/indirector/catalog/rest.rb#L14-L38)

There is also a [`post_catalog4`](https://www.rubydoc.info/gems/puppet/Puppet/HTTP/Service/Compiler:post_catalog4)

Reply:
Thanks for the note. I will keep that in mind for further refactoring.